Userland/unzip.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2020, Andrés Vieira <anvieiravazquez@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <string.h>
+#include <sys/stat.h>
+
+static const u8 central_directory_file_header_sig[] = "\x50\x4b\x01\x02";
+
+bool seek_and_read(u8* buffer, Core::File& file, off_t seek_to, int number_of_reads)
+{
+    if (!buffer)
+        return false;
+
+    if (!file.seek(seek_to))
+        return false;
+
+    if (file.read(buffer, number_of_reads) != number_of_reads)
+        return false;
```

Comment:
You should instead loop while the total number of bytes read is less than we need (checking for errors).

Reply:
I don't understand that. Isn't Core::IODevice::read(...) supposed to return the number of bytes that it read? So if it reads something that is not exactly the size we need it's wrong data.

I may be wrong, also I don't get how the while loop could look like, sorry :(

Reply:
> So if it reads something that is not exactly the size we need it's wrong data.

Not necessarily wrong data, but maybe it's not all of the data we want. A while loop could look like this:

```cpp
size_t need_bytes = ...;
size_t read_bytes = 0;

while (read_bytes < need_bytes) {
    auto nread = file.read(buffer + read_bytes, need_bytes - read_bytes);
    if (nread <= 0)
        return false;
    read_bytes += nread;
}
```

Reply:
I see. I assumed read() always read the number of bytes provided. Fixed it. Thank you so much :)